#167729785 Signup email verification

mthamayor opened the pull request:
#### What does this PR do?
Users can receive verification email during signup

#### Description of Task to be completed?
1. Create user signup endpoint POST /api/v1/auth/signup
2. Create verification endpoint PUT /api/v1/auth/verify/:token

#### How should this be manually tested?
- Install dependencies with `npm install`
- Startup the localhost server with `npm run dev`
- On postman, Access the post route with the following body
  
![Screenshot (178)](https://user-images.githubusercontent.com/16121915/63558099-6df89880-c543-11e9-938c-b5986f596bdb.png)
- Hit send, and you should get a response similar to a sample below
![Screenshot (179)](https://user-images.githubusercontent.com/16121915/63558158-a8623580-c543-11e9-897e-de23d4d17fed.png)
- You should receive a verification email similar to this
![Screenshot (180)](https://user-images.githubusercontent.com/16121915/63558242-07c04580-c544-11e9-99ed-5302bf906dc8.png)

#### Any background context you want to provide?
- Moved documentation to public/api-docs as shown in the [simulations document](https://docs.google.com/document/d/1gaagRK8SpR7LM-5j9zuvUjq7mOFKle-0YyKsEjD5E84/edit) folder structure
- Added a middleware to check if the user had been validated
- Added `SENDGRID_API_KEY` and `WEB_MAIL_URL`, EMAIL_VERIFY_URL_TEST, EMAIL_VERIFY_URL_DEV, EMAIL_VERIFY_URL_PROD variables to your env file. Please contact me to get it.
- https://github.com/andela/shadowcat-backend/blob/b8800aab68082382953b9b9c650d54d75116b967/src/database/migrations/20190821140038-create-user.js#L27-L30
The allowNull property was set to true as requested by @sebaztyn as this allows him to store user details gotten from Google and Facebook. The details at his disposal do not have users' passwords. 

#### What are the relevant pivotal tracker stories?
https://www.pivotaltracker.com/story/show/167729785

#### Screenshots (if appropriate)
None

#### Questions:
None

mthamayor committed:
feature(signup-endpoint):create folder structure

mthamayor committed:
feature(signup-endpoint):set up unit test

mthamayor committed:
feature(signup-endpoint): remove sequelize integration

mthamayor committed:
fix(merge-conflict): resolve merge conflict

mthamayor committed:
fixes(merge-conflicts):fix merge conflicts

mthamayor committed:
feature(email-verification): implement sendGrid API

create signup endpoint
integrate sendGrid api
create email verification endpoint

mthamayor committed:
feature(email-verification): implement sendGrid api

code cleanup

[Finishes #167729785]

mthamayor committed:
fixes(merge-conflict):resolve merge conflict

mthamayor committed:
refactor(email-verify):remove redundant field from env

sebaztyn commented on src/database/migrations/20190821140038-create-user.js:
```diff
@@ -32,5 +49,5 @@ export default {
     }
   }),
-  down: queryInterface => queryInterface.dropTable('Users')
+  down: (queryInterface) => queryInterface.dropTable('Users')
```
@mthamayor kindly include an eslint ignore script within your eslintrc file...to help ignore this and commit again please

mthamayor replied:
Okay. Thank you. I'll do just that

mthamayor replied:
@sebaztyn This has been resolved. Thanks.

mthamayor committed:
fixes(hound):attempt hound warning fix

Stylll commented on src/controllers/validate.js:
```diff
@@ -0,0 +1,26 @@
+      { active: true },
+      { returning: true, where: { userId: validateEntry.userId } }
+    );
+    res.status(201).json({ status: 201, data: 'Registration complete' });
```
I don't think this should be a 201 status since the resource is just being updated, not created.

mthamayor replied:
Okay. I will work on that

Stylll commented on src/database/migrations/20190821140038-create-user.js:
```diff
@@ -7,21 +7,38 @@ export default {
     password: {
       type: Sequelize.STRING,
-      allowNull: false,
+      allowNull: true
```
Is there a reason we are allowing nulls for the password field?

mthamayor replied:
@Stylll This was requested by Chima for the social media verification

Stylll commented on src/models/validate.js:
```diff
@@ -0,0 +1,14 @@
+
+export default (sequelize, DataTypes) => {
+  const Validate = sequelize.define('Validate', {
+    userId: DataTypes.INTEGER,
```
I think the data type here should be UUID since that's the datatype you set in the validate migration file is also UUID.

mthamayor replied:
Okay. I will look into this.

Stylll commented on src/routes/api/auth.js:
```diff
@@ -0,0 +1,11 @@
+
+authRouter.post('/signup', signupValidator, signupController);
+
+authRouter.get('/verify/:token', validateController);
```
I feel this is more of a `PUT` request since the user resource is actually being updated before it is returned.

mthamayor replied:
Thanks. I will update accordingly.

Stylll commented on src/services/sendGrid.js:
```diff
@@ -0,0 +1,44 @@
+  } else if (process.env.NODE_ENV === 'test') {
+    verificationUrl = 'http://localhost:3001/api/v1/auth/verify/';
+  } else {
+    verificationUrl = 'https://shadowcat-backend-staging.herokuapp.com';
```
We can put these `verificationUrl`s in the env file so if they need to be changed later, we won't need to update the codebase.

mthamayor replied:
Okay. I will do that.

Stylll commented on src/utils/constants/errorMessages.js:
```diff
@@ -0,0 +1,16 @@
+  undefinedPhone: 'Phone number is required',
+  undefinedPassword: 'Password is required',
+  invalidPhone: 'Phone number not valid',
+  invalidPassword: 'Password cannot be less then 8 characters',
```
`than` not `then`.

mthamayor replied:
Thank you for pointing it out

Stylll commented on src/validation/signup.js:
```diff
@@ -0,0 +1,66 @@
+    .withMessage(signupErrors.undefinedEmail)
+    .isEmail()
+    .withMessage(signupErrors.invalidEmail)
+    .matches(/andela.com$/)
```
I believe this won't catch an email like this `james@bondandela.com`. You can update the regex to `/@andela.com$/` instead

mthamayor replied:
Thank you. I will update accordingly.

Stylll reviewed (changes requested):
Hey @mthamayor Good job so far. Just a few updates. Kindly review my comments.

mthamayor committed:
feature(email-verify):implement PR changes

mthamayor committed:
feature(email-verify):implement PR changes

Stylll reviewed (changes requested):
@mthamayor so I noticed the error messages are returned one by one. for example, If I don't pass anything in the body, the response says to enter firstname, then after entering firstname, the response says to enter Lastname. It'll be nice if the response includes all the failed validation so the number of request to the API is reduced and the user can know all the fields to fill and submit.

mthamayor committed:
feature(email-verify):return array of all errors

mthamayor committed:
feature(email-verify):refactor verification endpoint

write swagger documentation for signup and email verification endpoints
code cleanup

chibytez reviewed (approved):
nice work, keep it up

mthamayor committed:
feature(signup-validate): validate signup form

Stylll reviewed (changes requested):
I noticed the error message is in this format when no email or first name is passed.
```
{
    "status": 400,
    "errors": [
        "First Name is required",
        "First name can only be alphabets",
        "Email is required",
        "Email provided is not valid",
        "Andela email not provided",
        "Password is required",
        "Password cannot be less than 8 characters",
        "Password should contain no special characters with at least one digit and one letter",
        "Phone number is required",
        "Phone number not valid"
    ]
}
```

Can we have it in a way where it lets the user know what fields are having the error

```
{
    "status": 422,
    "errors": {
       "email": "Input a valid email address",
       "password": "password is required"
    }
}
```

OR

```
{
    "status": 422,
    "errors": {
       "email": ["Input a valid email address", "email should be an andela email"],
       "password": ["password is required"]
    }
}
```

mthamayor committed:
feature(email-verify): make error message more descriptive

mthamayor committed:
feature(email-verify): make error message more descriptive

Stylll reviewed (changes requested):
@mthamayor Good job so far. Just a few more things

1. You should update your test script so it cleans up the database and run migrations before running tests. You can reach out to @chibytez or check out his PR, he has a script that does that.
2. I noticed the phone number length is not validated. We should only allow an 11 digit number.
3. For the password, let's allow atleast one special character, so a password must look like something like this `Stephen%2`

mthamayor committed:
feature(email-verify):implement phone length validation

resolve migration issues
add phone length test

mthamayor committed:
feature(password-validate):add at least one special character

okafor-chidimma reviewed (approved):
Great job you have done here @mthamayor All tests cases are passing

mthamayor committed:
feature(email-verify): add success message

update .env.sample to include sendgrid parameter
update test file

mthamayor committed:
feature(email-verify):format route index

mthamayor committed:
feature(email-verify):resolve merge conflicts

mthamayor committed:
feature(email-verify):fix syntax on constants

mthamayor committed:
feature(email-verify):refactor routes

okabamac reviewed (approved):
All tests are passing and it looks good to me. Well done!
